cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/store/DefaultMetadataStore.java

```diff
@@ -345,6 +349,12 @@ public void removeMetadata(final MetadataScope scope, final MetadataEntity metad
                                                      metadataChange.getExisting().getProperties(),
                                                      metadataChange.getExisting().getTags());
     publishAudit(previous, new MetadataRecordV2(metadataEntity, scope), new MetadataRecordV2(previous));
+    // delete all sub entities metadata
+    Set<MetadataSearchResultRecordV2> subEntities = getSubEntities(metadataEntity);
+    for (MetadataSearchResultRecordV2 subEntity : subEntities) {
+      LOG.trace("Deleting sub-entity {} metadata", subEntity.getMetadataEntity());
+      removeMetadata(subEntity.getMetadataEntity());
+    }
```

Comment:
who removes the parent entity index entries?

Reply:
if an entity has N sub-entities, then this will result in how many transactions? Looks like it will be 3 transactions for each entity involved (one for remove in each scope, one for sub-entity search). So it is (N+1 * 3). For 1000 sub-entities, this is 3003 transactions? Wow!

Reply:
You could reduce this by only searching for sub-entities if it is a known entity.

Reply:
The getSubEntities() in its current implementation only performs a search for known entity if its custom it just returns an empty result set. Added doc to document the behavior.

Did the change which you suggested in person to only call remove in user scope saving some more transactions.